### PR TITLE
chore: rework of CookieThemeProvider

### PR DESCRIPTION
**Pull request by Jabher**

### CODE REVIEW

File: packages/theme/src/cookie-theme-provider.tsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    overrideTheme?: ThemeName
+  }>
+> = React.memo(
+  ({
+    children,
+    initialTheme,
+    // overrideTheme is mainly used for storybook
+    overrideTheme,
+  }) => {
+    const parentTheme = useContext(ThemeToggleContext)
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/cookie-theme-provider.tsx:47]: Why do we import theme from ThemeToggle? From the name, I can only assume it's only toggle theme, but not export it

File: packages/theme/src/cookie-theme-provider.tsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  PropsWithChildren<{
+    // Use themeNameParent if you need get cookie in SSR
+    initialTheme?: ThemeName
+    overrideTheme?: ThemeName
+  }>
+> = React.memo(
+  ({
+    children,
+    initialTheme,
+    // overrideTheme is mainly used for storybook
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/cookie-theme-provider.tsx:44]: Thanks for description!

File: packages/theme/src/cookie-theme-provider.tsx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      initialTheme || DEFAULT_THEME_NAME
+    )
+    const themeName = parentTheme.themeName || internalThemeName
+
+    const theme = themeMap[themeName]
+
+    // Noticing browser preferences on hydration
+    // Reacting to changing preferences
+    useEffect(() => {
+      if (typeof window === 'undefined' || !isTopLevelProvider) {
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/cookie-theme-provider.tsx:59]: Next doesn't run effects on server, so it's safe to omit `typeof window === 'undefined'`

File: packages/theme/src/cookie-theme-provider.tsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+      const setTheme = () => {
+        const systemThemeName = prefersDarkThemeMediaQuery?.matches
+          ? ThemeName.dark
+          : ThemeName.light
+        const themeNameCookie = getThemeNameFromCookies()
+        const newThemeName =
+          overrideTheme ||
+          themeNameCookie ||
+          initialTheme ||
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/cookie-theme-provider.tsx:71]: Why not use internalThemeName? What is internalTheme? I'm still a bit confused, why does it go before systemTheme?

File: packages/theme/src/document-head-contents/element-theme-colors.tsx [hunk 5]
```diff
@@ -1,34 +1,35 @@
...
+color-theme: light;
+${lightThemeColors}
 }
 @media (prefers-color-scheme: dark) {
-    html:not([data-lido-theme='${ThemeName.light}']) { 
-        color-theme: dark;
-        ${darkThemeColors} 
-    }
+html:not([data-lido-theme='${ThemeName.light}']) { 
+color-theme: dark;
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/document-head-contents/element-theme-colors.tsx:18]: Missing spaces?

File: packages/theme/src/theme-provider.stories.tsx [hunk 6]
```diff
@@ -20,20 +21,34 @@ const themeCustom = {
   },
 }
 
+const Block: FC<PropsWithChildren> = styled.div`
+  font-size: ${({ theme }) => theme.fontSizesMap.xs}px;
+  line-height: 1.5em;
+  padding: ${({ theme }) => theme.spaceMap.xxl}px;
+  margin: ${({ theme }) => theme.spaceMap.md}px 0;
+  border-radius: ${({ theme }) => theme.borderRadiusesMap.xl}px;
+  background: ${({ theme }) => theme.colors.foreground};
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/theme-provider.stories.tsx:30]: Shouldn't we use css variables here?

File: packages/theme/src/theme-provider.tsx [hunk 7]
```diff
@@ -0,0 +1,46 @@
+import React, { FC, PropsWithChildren } from 'react'
+import styled, { ThemeProvider as StyledThemeProvider } from 'styled-components'
+import { reverseThemeMap, themeDark, themeDefault, themeLight } from './themes'
+import { initColors } from './document-head-contents'
+import { generateCssColorVariables } from './utils/generate-css-color-variables'
+import { Theme } from './types'
+
+// you need this if you're using themes, so we initiate this automatically
+initColors()
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/theme-provider.tsx:9]: Why do we call it here and in cookied provider?

File: packages/theme/src/utils/cookies.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import { ThemeName } from '@lidofinance/theme'
+import { themeCookieKey } from '../constants'
+
+const cookieThemeMatcher = new RegExp(`(^| )${themeCookieKey}=([^;]+)`)
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/utils/cookies.ts:4]: Maybe we should mention that we have similar regex in [element-theme-script.tsx](https://github.com/lidofinance/ui/pull/350/files#diff-e8dbc3eda19e57ac79a8698d709af83bea63dd9d3071cedf2b661533a94bce72)?

**Review by kozlovzxc-at-lido — APPROVED** [review]

File: packages/theme/src/theme-provider.stories.tsx (hunk 6 above)
Comment (kozlovzxc-at-lido) [reply packages/theme/src/theme-provider.stories.tsx:30]: Maybe we should delete colors from theme?

File: packages/theme/src/cookie-theme-provider.tsx (hunk 1 above)
Comment (Jabher) [reply packages/theme/src/cookie-theme-provider.tsx:47]: this is what we're providing in this component, so if something is provided, it's nested

File: packages/theme/src/document-head-contents/element-theme-colors.tsx (hunk 5 above)
Comment (Jabher) [reply packages/theme/src/document-head-contents/element-theme-colors.tsx:18]: smaller builds

File: packages/theme/src/theme-provider.tsx (hunk 7 above)
Comment (Jabher) [reply packages/theme/src/theme-provider.tsx:9]: if we're using any of providers this should be ran before react. will fix

**Pull request merged**